pkg/backend/filestate/backend.go

```diff
@@ -233,19 +237,102 @@ func New(ctx context.Context, d diag.Sink, originalURL string, project *workspac
 	}
 	backend.currentProject.Store(project)
 
+	projectMode := true
 	switch v := pulumiState.Version; v {
 	case 0:
 		backend.store = newLegacyReferenceStore(wbucket)
+		projectMode = false
 	case 1:
 		backend.store = newProjectReferenceStore(wbucket, backend.currentProject.Load)
 	default:
 		return nil, fmt.Errorf(
 			"state store unsupported: 'Pulumi.yaml' version (%d) is not supported "+
 				"by this version of the Pulumi CLI", v)
 	}
+
+	if !projectMode {
+		return backend, nil
+	}
+
+	// If we're in project mode warn about any old stack files.
+	refs, err := newLegacyReferenceStore(wbucket).ListReferences()
+	if err != nil {
+		// If there's an error listing don't fail, just don't print the warnings
+		return backend, nil
+	}
+
+	for _, ref := range refs {
+		d.Warningf(&diag.Diag{
+			Message: "Found legacy stack file '%s', you should run 'pulumi state upgrade'",
+		}, ref.Name())
+	}
+
 	return backend, nil
 }
 
+func (b *localBackend) Upgrade(ctx context.Context) error {
+	// We don't use the existing b.store because
+	// this may already be a projectReferenceStore
+	// with new legacy files introduced to it accidentally.
+	olds, err := newLegacyReferenceStore(b.bucket).ListReferences()
+	if err != nil {
+		return fmt.Errorf("read old references: %w", err)
+	}
+
+	newStore := newProjectReferenceStore(b.bucket, b.currentProject.Load)
+	var upgraded int
+	for _, old := range olds {
+		if err := b.upgradeStack(ctx, newStore, old); err != nil {
+			b.d.Warningf(diag.Message("", "Skipping stack %q: %v"), old, err)
+			continue
+		}
+		upgraded++
+	}
```

Comment:
If it's remote blob storage, is there anything keeping us from running this loop in parallel?

Reply:
Yeah, I think we could do this concurrently. I'll make that change.

Reply:
Concurrency added. Had to make diag.Sink thread-safe. #12485